# KGD
**K**inship (genetic relatedness) using **G**BS (genotyping-by-sequencing) with **D**epth adjustment

R code for the analysis of genotyping-by-sequencing (GBS) data, primarily to construct a genomic relationship matrix for the genotyped individuals. The code can be used on its own, or incorporated into other R programs. There are QC tools (primarily graphical output), relationship estimation tools, pedigree verification tools and pedigree ‘mix and match’ tools. The latter two operations require additional input information about the samples genotyped.

The manual describing the use of the code is **KGDManual.pdf** in the top directory of this repository.

This project was supported by the Ministry of Business, Innovation and Employment via its funding of the “Genomics for Production & Security in a Biological Economy” programme (Contract ID C10X1306). 

Please cite

Dodds, K G, McEwan, J C, Brauning, R, Anderson, R A, Van Stijn, T C, Kristjánsson, T and Clarke, S M (2015) Construction of relatedness matrices using genotyping-by-sequencing data. *BMC Genomics* **16**, 1047. [DOI](https://doi.org/10.1186/s12864-015-2252-3) [GoogleScholar citations](https://scholar.google.com/scholar?hl=en&cites=10840046335958418721)

for parentage assignment

Dodds, K G, McEwan, J C, Brauning, R, Van Stijn, T C, Rowe, S J, McEwan, K M and Clarke, S M (2019) Exclusion and genomic relatedness methods for assignment of parentage using genotyping-by-sequencing data. *G3: Genes, Genomes, Genetics* **9**, 3239-3247. [DOI](https://doi.org/10.1534/g3.119.400501) [GoogleScholar citations](https://scholar.google.com/scholar?hl=en&cites=16774971615445565137)

for population differences (Fst functions)

Dodds, K G, McEwan, J C, Brauning, R, & Clarke, S M (2025). Assessing population allele frequency differences using low-depth sequencing data. Journal of the Royal Society of New Zealand, (early online). [DOI] (https://doi.org/10.1080/03036758.2025.2500999)
